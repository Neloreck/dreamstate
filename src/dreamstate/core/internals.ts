import type { Context } from "react";

import type {
  TAnyContextManagerConstructor,
  TAnyValue,
  TQuerySubscriptionMetadata,
  TSignalSubscriptionMetadata,
} from "@/dreamstate/types";

/**
 * Meta symbols for private internals in context managers.
 */
export const SIGNAL_METADATA_SYMBOL: unique symbol = Symbol("SIGNAL_METADATA");
export const QUERY_METADATA_SYMBOL: unique symbol = Symbol("QUERY_METADATA");
export const SIGNALING_HANDLER_SYMBOL: unique symbol = Symbol("SIGNALING_HANDLER");
export const SCOPE_SYMBOL: unique symbol = Symbol("SCOPE");

/**
 * A weak map registry that stores React context instances bound to specific manager classes.
 *
 * This registry ensures that the library does not retain unnecessary references to manager classes,
 * preventing memory leaks and unintended side effects.
 *
 * This is particularly useful in scenarios such as:
 * - Hot Module Replacement
 * - Module unloading
 * - Scope disposal
 */
export const CONTEXT_REACT_CONTEXTS_REGISTRY: WeakMap<
  TAnyContextManagerConstructor,
  Context<TAnyValue>
> = new WeakMap();

export const SIGNAL_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TSignalSubscriptionMetadata> =
  new WeakMap();

export const QUERY_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TQuerySubscriptionMetadata> =
  new WeakMap();
