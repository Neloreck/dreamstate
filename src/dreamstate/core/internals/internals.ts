import type { Context } from "react";

import type {
  TServiceMap,
  TQuerySubscriptionMetadata,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TUpdateObserver,
  TUpdateSubscriber,
  TDreamstateService,
  TAnyObject
} from "@/dreamstate/types";

/**
 * Global shared constants.
 */

export const EMPTY_ARR: [] = [];

/**
 * Store related constants.
 */

// Currently active services instance types.
export const CONTEXT_SERVICES_ACTIVATED: Set<TDreamstateService> = new Set();

// Currently active services instance types.
export const CONTEXT_SERVICES_REFERENCES: TServiceMap<number> = new WeakMap();

// ReactContext registry, lazy initialized constants.
export const CONTEXT_REACT_CONTEXTS_REGISTRY: TServiceMap<Context<any>> = new WeakMap();

// ContextService instances registry.
export const CONTEXT_SERVICES_REGISTRY: TServiceMap<InstanceType<TDreamstateService>> = new WeakMap();

// ContextManagers context references registry.
export const CONTEXT_STATES_REGISTRY: TServiceMap<TAnyObject> = new WeakMap();

// ContextManagers observers registry - Providers of related context.
export const CONTEXT_OBSERVERS_REGISTRY: TServiceMap<Set<TUpdateObserver>> = new WeakMap();

// ContextManagers subscribers registry for data updates.
export const CONTEXT_SUBSCRIBERS_REGISTRY: TServiceMap<Set<TUpdateSubscriber<any>>> = new WeakMap();

// Signal handlers registry bound to context instances.
export const CONTEXT_SIGNAL_HANDLERS_REGISTRY: TServiceMap<TSignalListener<any>> = new WeakMap();

// Signal listeners metadata registry related to services.
export const CONTEXT_SIGNAL_METADATA_REGISTRY: TServiceMap<TSignalSubscriptionMetadata> = new WeakMap();

// Query listeners metadata registry related to services.
export const CONTEXT_QUERY_METADATA_REGISTRY: TServiceMap<TQuerySubscriptionMetadata> = new WeakMap();

// Global signal listeners registry.
export const SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>> = new Set();
