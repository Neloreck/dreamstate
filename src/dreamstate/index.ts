/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export {
  ILoadable as Loadable,
  TNested as Nested,
  TComputed as Computed,
  TStateSetter as StateSetter,
  TConsumable as Consumable,
  ISignalEvent as SignalEvent,
  ISignal as Signal,
  TSignalType as SignalType,
  TQueryResponse as QueryResponse,
  TOptionalQueryResponse as OptionalQueryResponse,
  TQueryRequest as QueryRequest,
  TQueryType as QueryType
} from "@/dreamstate/types";

export { ContextService } from "@/dreamstate/core/services/ContextService";
export { ContextManager } from "@/dreamstate/core/services/ContextManager";

export { getCurrentContext } from "@/dreamstate/core/registry/getCurrentContext";
export { getCurrent } from "@/dreamstate/core/registry/getCurrent";

export { createProvider } from "@/dreamstate/core/provision/createProvider";
export { Provide } from "@/dreamstate/core/provision/Provide";
export { withProvision } from "@/dreamstate/core/provision/withProvision";

export { OnSignal } from "@/dreamstate/core/signals/OnSignal";
export { useSignals } from "@/dreamstate/core/signals/useSignals";
export { unsubscribeFromSignals } from "@/dreamstate/core/signals/unsubscribeFromSignals";
export { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
export { emitSignal } from "@/dreamstate/core/signals/emitSignal";

export { OnQuery } from "@/dreamstate/core/queries/OnQuery";
export { queryData } from "@/dreamstate/core/queries/queryData";

export { useManager } from "@/dreamstate/core/consumption/useManager";
export { Consume } from "@/dreamstate/core/consumption/Consume";
export { withConsumption } from "@/dreamstate/core/consumption/withConsumption";

export { Bind } from "@/dreamstate/utils/Bind";
export { createLoadable } from "@/dreamstate/utils/createLoadable";
export { createNested } from "@/dreamstate/utils/createNested";
export { createComputed } from "@/dreamstate/utils/createComputed";
export { createSetter } from "@/dreamstate/utils/createSetter";
