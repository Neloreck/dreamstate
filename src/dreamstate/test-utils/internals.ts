/**
 * Meta symbols for private internals in context managers for test-utils.
 */
export const IS_MOCKED: unique symbol = Symbol("IS_MOCKED");
