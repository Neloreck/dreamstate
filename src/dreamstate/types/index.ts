export * from "@/dreamstate/types/queries";
export * from "@/dreamstate/types/signals";
export * from "@/dreamstate/types/error";
export * from "@/dreamstate/types/general";
export * from "@/dreamstate/types/decorators";
export * from "@/dreamstate/types/utils";
export * from "@/dreamstate/types/internal";
