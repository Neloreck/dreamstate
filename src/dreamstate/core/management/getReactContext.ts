import { Context, createContext } from "react";

import { CONTEXT_REACT_CONTEXTS_REGISTRY } from "@/dreamstate/core/internals";
import { IContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Retrieves the React context reference associated with the provided `ManagerClass`.
 * The context is lazily created only after the first access attempt. If no manager is provided in the scope,
 * a default context value can be applied.
 *
 * This function allows for the dynamic creation and retrieval of a context specific to a given manager,
 * ensuring that the context is properly initialized and available for use.
 *
 * @template S - The type of the context state.
 * @template M - The type of the context manager constructor.
 * @param {M} ManagerClass - The context manager constructor reference used to identify the context.
 * @returns {Context<S>} - A React context instance with a pre-defined default value.
 */
export function getReactContext<S extends TAnyObject, M extends IContextManagerConstructor<S>>(
  ManagerClass: M
): Context<S> {
  if (CONTEXT_REACT_CONTEXTS_REGISTRY.has(ManagerClass)) {
    return CONTEXT_REACT_CONTEXTS_REGISTRY.get(ManagerClass) as Context<S>;
  } else {
    const reactContext: Context<S> = createContext(ManagerClass.getDefaultContext() as S);

    /*
     * Later providers and consumers in tree will be displayed as
     * 'DS.Class.Provider' or 'DS.Class.Consumer'.
     */
    reactContext.displayName = "DS." + ManagerClass.name;

    CONTEXT_REACT_CONTEXTS_REGISTRY.set(ManagerClass, reactContext);

    return reactContext;
  }
}
