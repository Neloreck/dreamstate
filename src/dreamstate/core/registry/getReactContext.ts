import { Context, createContext } from "react";

import { debug } from "@/macroses/debug.macro";

import { CONTEXT_REACT_CONTEXTS_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get manager react context internal.
 */
export function getReactContext<T extends TAnyContextManagerConstructor>(
  Manager: T
): Context<T> {
  debug.info("Requested manager react context:", Manager.name);

  if (CONTEXT_REACT_CONTEXTS_REGISTRY.has(Manager)) {
    return CONTEXT_REACT_CONTEXTS_REGISTRY.get(Manager)!;
  } else {
    const reactContext: Context<T> = createContext(null as any);

    reactContext.displayName = "DS." + Manager.name;

    CONTEXT_REACT_CONTEXTS_REGISTRY.set(Manager, reactContext);

    debug.info("Context manager context declared:", Manager.name, reactContext.displayName);

    return reactContext;
  }
}
