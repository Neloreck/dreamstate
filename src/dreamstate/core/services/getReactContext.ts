import { Context, createContext } from "react";

import { CONTEXT_REACT_CONTEXTS_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get react context reference for provided ManagerClass.
 * Lazy internal creation only after first access attempt.
 */
export function getReactContext<
  T extends TAnyContextManagerConstructor
>(
  ManagerClass: T
): Context<T> {
  if (CONTEXT_REACT_CONTEXTS_REGISTRY.has(ManagerClass)) {
    return CONTEXT_REACT_CONTEXTS_REGISTRY.get(ManagerClass) as Context<T>;
  } else {
    const reactContext: Context<T> = createContext(null as any);

    /**
     * Later providers and consumers in tree will be displayed as
     * 'DS.Class.Provider' or 'DS.Class.Consumer'.
     */
    reactContext.displayName = "DS." + ManagerClass.name;

    CONTEXT_REACT_CONTEXTS_REGISTRY.set(ManagerClass, reactContext);

    return reactContext;
  }
}
