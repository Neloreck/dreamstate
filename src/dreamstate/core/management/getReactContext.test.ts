import { mount } from "enzyme";
import { Context, createElement, ReactElement } from "react";

import { ContextManager, useManager } from "@/dreamstate";
import { CONTEXT_REACT_CONTEXTS_REGISTRY } from "@/dreamstate/core/internals";
import { getReactContext } from "@/dreamstate/core/management/getReactContext";
import { TAnyObject } from "@/dreamstate/types";
import { ITestContext, TestManager } from "@/fixtures";

describe("getReactContext method functionality", () => {
  /**
   * Clear possible existing entry of manager for testing.
   */
  beforeEach(() => {
    CONTEXT_REACT_CONTEXTS_REGISTRY.delete(TestManager);
  });

  it("related react context should be lazily initialized correctly with changed displayName (dev)", () => {
    expect(CONTEXT_REACT_CONTEXTS_REGISTRY.get(TestManager)).toBeUndefined();

    const contextType: Context<TAnyObject> = TestManager.REACT_CONTEXT;

    expect(CONTEXT_REACT_CONTEXTS_REGISTRY.get(TestManager)).toBeDefined();

    expect(contextType).toBeDefined();
    expect(contextType.Consumer).toBeDefined();
    expect(contextType.Provider).toBeDefined();
    expect(contextType.displayName).toBe("Dreamstate." + TestManager.name);

    CONTEXT_REACT_CONTEXTS_REGISTRY.delete(TestManager);
  });

  it("related react context should be lazily initialized correctly with changed displayName (dev)", () => {
    IS_DEV = false;
    expect(TestManager.REACT_CONTEXT.displayName).toBeUndefined();
  });

  it("should supply null value if manager is not provided", () => {
    function Consumer(): ReactElement {
      const context: ITestContext = useManager(TestManager);

      return createElement("div", {}, JSON.stringify(context));
    }

    expect(mount(createElement(Consumer)).render()).toMatchSnapshot();
  });

  it("should supply default value if manager is not provided", () => {
    class WithDefaultsManager extends ContextManager<TAnyObject> {
      public static getDefaultContext(): Partial<TAnyObject> {
        return {
          first: "unknown",
          second: -1,
        };
      }

      public readonly context: TAnyObject = {
        first: "first",
        second: 2,
        third: false,
      };
    }

    getReactContext(WithDefaultsManager);

    function Consumer(): ReactElement {
      const context: TAnyObject = useManager(WithDefaultsManager);

      return createElement("div", {}, JSON.stringify(context));
    }

    expect(mount(createElement(Consumer)).render()).toMatchSnapshot();
  });
});
