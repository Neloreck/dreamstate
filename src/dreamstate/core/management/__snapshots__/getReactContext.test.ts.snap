// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReactContext method functionality should supply default value if manager is not provided 1`] = `
<div>
  {"first":"unknown","second":-1}
</div>
`;

exports[`getReactContext method functionality should supply null value if manager is not provided 1`] = `
<div>
  null
</div>
`;
